Core authentication

A248 opened the pull request:
We will need to review the code paths closely, as well as add integration tests for them.

A248 committed:
Initial draft of mod-authentication

A248 committed:
SaltGenerator#emptySalt typo

auriium reviewed (approved):
Everything looks good! We should probably start doing misc-changes like the SimpleDataKeySPI on the main branch however, to allow us to simply rebase onto master.

auriium commented on mod-authentication/src/main/java/gg/solarmc/loader/authentication/AuthManager.java:
```diff
@@ -0,0 +1,221 @@
+	/**
+	 * Assuming the user has entered the password correctly, complete their login. <br>
+	 * <br>
+	 * Will migrate the user's previous data if they are now premium, but used to be cracked
```
I like this! Although, i initially thought there would be a manual way of migrating accounts rather than automatically migrating if a premium user logged in, that way you could choose if you wanted to remain cracked or not.

A248 replied:
Yeah, I doubt anyone would want to remain cracked. There don't seem to be any advantages to it.

auriium reviewed (commented):
Marked all files as reviewed- waiting to see future changes to this draft!

A248 committed:
Rename mod-auth to core-authentication

A248 committed:
Make AuthenticationCenter standalone object; add login callbacks

* UnauthenticatedUser must be implemented by the platform in
  conjunction with DataLoader. The AuthenticationCenter will
  trigger the loading of new or existing data depending on what
  happens with respect to adding the user.
* The downstream auth plugin must be careful to check the result
  of all AuthenticationCenter methods.
* In one area of the implementation, JDBC had to be directly used,
  so that is now noted in the schema file.

A248 committed:
Fix compilation of module-info

A248 committed:
Fix string too large test failure

auriium commented on core-authentication/src/main/java/gg/solarmc/loader/authentication/UnauthenticatedUser.java:
```diff
@@ -19,51 +19,43 @@
+import gg.solarmc.loader.Transaction;
 
-	public UnauthenticatedUser(UUID mcUuid, String username) {
-		this.mcUuid = Objects.requireNonNull(mcUuid);
```
This isnt really a proper review but it caught my eye - Shouldn't we use some kind of standardized @Nullable annotation instead of adding a line every time we want to verify a variable for nullity?

A248 replied:
Nullability annotations are not enforced at runtime. Some of them (with Retention(RUNTIME)) are enforceable if you use AOP and write a brief runtime nullness checker, but I do not think we need to do that. Objects.requireNonNull is perfectly suitable

auriium commented:
Looks good to merge!

A248 committed:
Merge branch 'master' into core-authentication

A248 committed:
Refactor SQL migrations and core auth; Use routines for auth

* Split the SQL migrations for each data module into their own
  file. Avoids a crowded v1 file.
* Use stored functions to re-organize some of the SQL in
  core-authentication and reduce network traffic.

auriium commented on core-authentication/src/main/java/gg/solarmc/loader/authentication/AuthenticationCenter.java:
```diff
@@ -20,20 +20,17 @@
 import java.util.UUID;
 import java.util.concurrent.ForkJoinPool;
 
+import static gg.solarmc.loader.schema.Routines.insertAutomaticAccountAndGetUserId;
```
I thought we agreed we weren't using routines table-side?

auriium commented on core-schema/src/main/resources/sql-schema/V5__Mod_credits.sql:
```diff
@@ -0,0 +1,11 @@
+  FOREIGN KEY (user_id) REFERENCES user_ids (id) ON DELETE CASCADE
+);
+
+CREATE VIEW credits_with_names AS
```
what is the point of this?

A248 replied:
It's used to display baltop info. I have found that selecting from views which join is somewhat more manageable than selecting joins. The database optimizes it to the same query plan anyway.

auriium reviewed (commented):
Nice work on this module! Looking closer and closer to merge!

A248 committed:
Remove unused code and update comment in AuthenticationCenter
